ckanext/dcatapchharvest/profiles.py

```diff
@@ -309,32 +322,41 @@ def _clean_end_datetime(self, datetime_value, data_type):
         Accepted types are: xsd:date, xsd:dateTime, xsd:gYear, or
         xsd:gYearMonth; or schema:Date or schema:DateTime, for temporals
         specified as schema:startDate and schema:endDate.
+
+        We only consider the parts of the date that are expected from the given
+        data_type, e.g. the year of an xsd:gYear, even if the month and day
+        have been included in the datetime_value. If a datetime_value with
+        data_type of xsd:dateTime or schema:DateTime does not contain time
+        information, we discard it.
         """
         try:
-            dt = isodate.parse_datetime(datetime_value)
-            if isinstance(dt, datetime):
-                # We already have a full datetime, no need to change it.
-                return datetime_value
-        except isodate.isoerror.ISO8601Error:
-            pass
+            if data_type == XSD.dateTime or data_type == SCHEMA.DateTime:
+                dt = isodate.parse_datetime(datetime_value)
 
-        try:
-            d = isodate.parse_date(datetime_value)
-            if data_type == XSD.date or data_type == SCHEMA.Date:
+                return dt.isoformat()
+            elif data_type == XSD.date or data_type == SCHEMA.Date:
+                dt = datetime.strptime(datetime_value, DATE_FORMAT)
                 end_datetime = datetime.max.replace(
-                    year=d.year, month=d.month, day=d.day)
+                    year=dt.year, month=dt.month, day=dt.day)
+
+                return end_datetime.isoformat()
```

Comment:
This part is pretty much the same that is happening up at line 293-310.
Wondering if it would be easier or harder to read, to put it in a function. At least it would be easier to maintain or spot errors if those functions diverge?

Reply:
thanks, I think that would be useful. The two methods were a lot more different before this little refactoring, so it made less sense to do that, but now I think it's worthwhile. :)

Reply:
... actually, I looked at the code again and I think that the current status is the best balance of readability and maintainability, at least that I can manage right now. :sweat_smile: Maybe we'll find a nicer way in the future, but for now I'd rather leave it this way.

Reply:
that's fine, thank you for double-checking :)